src/common/io/http/response.c

```diff
@@ -0,0 +1,412 @@
+OBJECT_DEFINE_GET(Code, const, HTTP_RESPONSE, unsigned int, code);
+OBJECT_DEFINE_GET(Header, const, HTTP_RESPONSE, const HttpHeader *, header);
+OBJECT_DEFINE_GET(Reason, const, HTTP_RESPONSE, const String *, reason);
+
+/***********************************************************************************************************************************
+When response is done close/reuse the connection
+***********************************************************************************************************************************/
+static void
+httpResponseDone(HttpResponse *this)
+{
+    FUNCTION_LOG_BEGIN(logLevelTrace);
+        FUNCTION_LOG_PARAM(HTTP_RESPONSE, this);
+    FUNCTION_LOG_END();
+
+    ASSERT(this != NULL);
+    ASSERT(this->session != NULL);
+
+    // If close was requested by the server then free the session
+    if (this->closeOnContentEof)
+    {
+        httpSessionFree(this->session);
+
+        // Only update the close stats after a successful response so it is not counted if there was an error/retry
+        httpClientStat.close++;
+    }
+    // Else return it to the client do it can be reused
+    else
+        httpSessionDone(this->session);
+
+    this->session = NULL;
+
+    FUNCTION_LOG_RETURN_VOID();
+}
+
+/***********************************************************************************************************************************
+Read content
+***********************************************************************************************************************************/
+static size_t
+httpResponseRead(THIS_VOID, Buffer *buffer, bool block)
+{
+    THIS(HttpResponse);
+
+    FUNCTION_LOG_BEGIN(logLevelTrace);
+        FUNCTION_LOG_PARAM(HTTP_RESPONSE, this);
+        FUNCTION_LOG_PARAM(BUFFER, buffer);
+        FUNCTION_LOG_PARAM(BOOL, block);
+    FUNCTION_LOG_END();
+
+    ASSERT(this != NULL);
+    ASSERT(buffer != NULL);
+    ASSERT(!bufFull(buffer));
+    ASSERT(this->contentEof || this->session != NULL);
+
+    // Read if EOF has not been reached
+    size_t actualBytes = 0;
+
+    if (!this->contentEof)
+    {
+        IoRead *rawRead = httpSessionIoRead(this->session);
+
+        // If close was requested and no content specified then the server may send content up until the eof
+        if (this->closeOnContentEof && !this->contentChunked && this->contentSize == 0)
+        {
+            ioRead(rawRead, buffer);
+            this->contentEof = ioReadEof(rawRead);
+        }
+        // Else read using specified encoding or size
+        else
+        {
+            do
+            {
+                // If chunked content and no content remaining
+                if (this->contentChunked && this->contentRemaining == 0)
+                {
+                    // Read length of next chunk
+                    MEM_CONTEXT_TEMP_BEGIN()
+                    {
+                        this->contentRemaining = cvtZToUInt64Base(strPtr(strTrim(ioReadLine(rawRead))), 16);
```

Comment:
What is this hardcoded 16 for?

Reply:
It means base 16, i.e. we are converting from a hex string to uint64. We could add a constant for BASE16, BASE10, BASE8, BASE2, etc. to `common/convert` if you like but I'd rather do that in a separate commit.